compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -974,30 +1340,69 @@ fn transform_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, options: TransformTyOptio
         }
 
         ty::Dynamic(predicates, _region, kind) => {
-            ty = Ty::new_dynamic(
-                tcx,
-                transform_predicates(tcx, predicates),
-                tcx.lifetimes.re_erased,
-                *kind,
-            );
-        }
-
-        ty::Alias(..) => {
-            ty = transform_ty(
-                tcx,
-                tcx.normalize_erasing_regions(ty::ParamEnv::reveal_all(), ty),
-                options,
-            );
+            if is_fn_trait(tcx, predicates) {
```

Comment:
`dyn Fn()` is not special. You can do fn dispatch on all trait objects that have `Fn` supertraits, and this code still SIGILLs on this branch:

```rust
trait MyCallable: Fn() {}
impl<T: Fn()> MyCallable for T {}

fn callable(x: &dyn MyCallable) { x(); }

fn main() {
    let capture = "world";
    callable(&|| println!("hello, {capture}"));
}
```

Reply:
This is a good example of what I explained above, where we don't have enough information at the call site to encode the label correctly (i.e., it encoded a call on a `&dyn MyCallable` trait object, but the label assigned to the closure is an `fn()`) so even if we encoded the closure fully as a closure, the labels woudn't match.

Reply:
Fixed it too.